Source/Particles/Deposition/CurrentDeposition.H

```diff
@@ -0,0 +1,223 @@
+    const Real invvol = dxi*dyi*dzi;
+#endif
+
+    const Real xmin = xyzmin[0];
+    const Real ymin = xyzmin[1];
+    const Real zmin = xyzmin[2];
+
+    Dim3 shape_lo;
+    compute_shape_factor_lo<depos_order>(shape_lo);
+    const Real clightsq = 1.0/PhysConst::c/PhysConst::c;
+
+    // Loop over particles and deposit into jx_arr, jy_arr and jz_arr
+    ParallelFor( np_to_depose,
+                 [=] AMREX_GPU_DEVICE (long ip) {
+                     // --- Get particle quantities
+                     const Real gaminv = 1.0/std::sqrt(1.0 + uxp[ip]*uxp[ip]*clightsq
+						       + uyp[ip]*uyp[ip]*clightsq
+						       + uzp[ip]*uzp[ip]*clightsq);
+                     const Real wq  = q*wp[ip];
+                     const Real vx  = uxp[ip]*gaminv;
+                     const Real vy  = uyp[ip]*gaminv;
+                     const Real vz  = uzp[ip]*gaminv;
+                     // wqx, wqy wqz are particle current in each direction 
+                     const Real wqx = wq*invvol*vx;
+                     const Real wqy = wq*invvol*vy;
+                     const Real wqz = wq*invvol*vz;
+
+                     // --- Compute shape factors
+                     // x direction
+                     // Get particle position after 1/2 push back in position
+                     const Real xmid = (Real) (xp[ip]-xmin)*dxi-dts2dx*vx;
+                     // Compute shape factors for node-centered quantities
+                     Real AMREX_RESTRICT sx [depos_order + 1];
+		     const int j  = compute_shape_factor<depos_order>(sx,  xmid);
+                     // Compute shape factors for cell-centered quantities
+                     Real AMREX_RESTRICT sx0[depos_order + 1];
+		     const int j0 = compute_shape_factor<depos_order>(sx0, xmid-stagger_shift);
+                     
+#if (AMREX_SPACEDIM == 3)
+                     // y direction
+                     const Real ymid= (Real) (yp[ip]-ymin)*dyi-dts2dy*vy;
+                     Real AMREX_RESTRICT sy [depos_order + 1];
+		     const int k  = compute_shape_factor<depos_order>(sy,  ymid);
+                     Real AMREX_RESTRICT sy0[depos_order + 1];
+		     const int k0 = compute_shape_factor<depos_order>(sy0, ymid-stagger_shift);
+#endif
+                     // z direction
+                     const Real zmid=(Real) (zp[ip]-zmin)*dzi-dts2dz*vz;
+                     Real AMREX_RESTRICT sz [depos_order + 1];
+		     const int l  = compute_shape_factor<depos_order>(sz,  zmid);
+                     Real AMREX_RESTRICT sz0[depos_order + 1];
+		     const int l0 = compute_shape_factor<depos_order>(sz0, zmid-stagger_shift);
+
+                     // Deposit current into jx_arr, jy_arr and jz_arr
+#if (AMREX_SPACEDIM == 2)
+                     for (int iz=0; iz<=depos_order; iz++){
+                         for (int ix=0; ix<=depos_order; ix++){
+                             amrex::Gpu::Atomic::Add(
+                                 &jx_arr(lo.x+j0+ix+shape_lo.x, 
```

Comment:
It seems to me that, if you define `j`, `j0`, etc. as the index of **leftmost cell** that a particle can touch (as opposed as, currently, the cell immediately to the left of the particle), then you do not need to `shape_lo` and the associated functions.

This would involve having `return j-1` in `compute_shape_factor <2>` and `compute_shape_factor <3>`.

Would that work, or am I missing something?